Refactoring of OBC interface to allow solution via NEVP

vetschn opened the pull request:
I added Beyn's method and the spectral solver using it that @almaeder has been working on. I also added some better test cases for the OBC.

Some things to still do:

- [ ] ~~Add other NEVP solvers~~ (Only adding a full solver and Beyn's method for now)
- [x] Test that the batching works
- [x] Test that everything works on GPU
- [x] Maybe refine the interface bit more
- [x] Think of a nice way to save and reuse the result from the last iteration.
- [x] Test for Memoizers
- [x] Tests for different formulas in the `Spectral` solver
- [ ] ~~Determine best way of doing batched inversion on GPU~~

This should close #14 and some related issues.

vetschn committed:
Initial refactoring of OBC interface to allow spectral solution

I added Beyn's method and the spectral solver using it that @almaeder has been working on. I also added some better test cases for the OBC.

vetschn committed:
CRAY MPICH does not provide mpi_memory_alloc_kinds, but is always GPU aware.

vetschn committed:
Linter problems

Also removed the superfluous test

vetschn committed:
GPU issues

The array was not on the damn device

vetschn committed:
Bugfix

Fixed a bug that occurs when adding/subtracting unstacked sparse matrices from dsbsparse matrices

vetschn committed:
Add an nevp solver based on linearization and tests

I also changed Beyn very slightly (no more warnings) and made some small changes in the NEVP interface. The test cases will need some more work.

vetschn committed:
Corrections to the spectral OBC

Now things work in batches and you have the option between different formulas for computing the surface green's function from the eigenvalues and eigenmodes. I changed the test cases to slightly more sensible ones.

vetschn committed:
Remove unneccessary test combinations.

When the global stack shapes are at the top level we get many unneeded test combinations. Move them into the scope in which they are needed.

vetschn committed:
Add tests for different sectioning strategies.

The "greedy" strategy was untested.

vetschn committed:
Implementation of an OBC Memoizer

I wrote a little wrapper class for OBC algorithms that does refinement iterations and checks OBC quality. Should lead to significant speedups.

vetschn committed:
Suppress RuntimeWarnings in the detection of reflected modes.

vetschn committed:
isort

vetschn committed:
Added Lyapunov solvers.

I implemented a Lyapunov solver interface + another Memoizer for this. I also added three different implementations: a vectorized one, an iterative doubling one from Alex Maeder and the one that was already present in the other version of quatrex.

Everything works on batched inputs and is tested with some (stabilized) random inputs.

vetschn committed:
Some smaller changes to the OBC interface & SanchoRubio

I made it so that it does not raise an Exception, but just gives a RuntimeWarning if things don't converge. I also renamed the Memoizer to be specific to the OBC.

vetschn committed:
Refactor Lyapunov and OBC classes to use Solver naming convention.

vetschn committed:
Fix GPU issues.

I fixed some bugs that occur due to missing functionality in CuPy and i had to shift some data between device and host.

vetschn commented:
The failing test is due to the choice of parameters for Beyn. I tried to tune it but it still fails like 25% of time. We really would need to come up with a better way of constructing NEVP with pre-selected eigenvalues... :(

vincent-maillou commented on src/qttools/datastructures/dsbcoo.py:
```diff
@@ -231,7 +231,7 @@ def __iadd__(self, other: "DSBSparse | sparse.sparray") -> "DSBCOO":
         if sparse.issparse(other):
             lil = other.tolil()
-            sparray_data = lil[self.rows, self.cols]
+            sparray_data = lil[self.rows, self.cols].toarray()
```
Does the in-place addition/substraction of 2 DSBSparse matrices densify the blocks being touched? Does it retain/modify the sparsity pattern of the blocks or assume that the sparsity patterns of the 2 matrices matches?

vetschn replied:
To clarify: This change concerns the addition of a `scipy.sparse.sparray` to the `DSBCOO`. This is used to construct the bare system matrices from the hamiltonian `sparray`, etc.

The sparsity of the `DSBCOO` is assumed to always be fixed, so here we also only extract the elements from the `sparray` that match, i.e. `(self.rows, self.cols)`. In `scipy` only the `lil_array` supports this fancy-indexing, but it returns a new `lil_array` (of shape `(1, len(self.rows))`), thus the call to `.toarray()`.

I will add a test for this over on the datastructures branch, it's not covered atm :cowboy_hat_face:.

The in-place operations of two `DSBCOO` matrices always operates on the data buffers directly, so nothing gets densified and sparsity pattern remain unchanged. For this the two matrices have to be commensurable:
https://github.com/vincent-maillou/qttools/blob/609f8ddc6ba061da3b0b748a374d77eed7acf720/src/qttools/datastructures/dsbcoo.py#L213-L228

Is that alright from your side?

vincent-maillou replied:
Yep sounds good! thanks for the explanations

vincent-maillou commented on src/qttools/lyapunov/vectorize.py:
Is that batched?

vetschn replied:
Yes, all the Lyapunov solvers are:
https://github.com/vincent-maillou/qttools/blob/609f8ddc6ba061da3b0b748a374d77eed7acf720/tests/lyapunov/test_lyapunov.py#L17-L27

vincent-maillou replied:
top

vincent-maillou commented on src/qttools/nevp/full.py:
numpy.linalg.npla, here the eig will always be performed on CPU (at least for now)?

vetschn replied:
Yep, see the D2H/H2D transfers here 
https://github.com/vincent-maillou/qttools/blob/609f8ddc6ba061da3b0b748a374d77eed7acf720/src/qttools/nevp/beyn.py#L97-L103

vincent-maillou replied:
Indeed, thanks!

vincent-maillou commented on src/qttools/nevp/utils.py:
```diff
@@ -0,0 +1,48 @@
+    return z, w
+
+
+def operator_inverse(
```
Need documentation for this function, don't know what it's doing

vetschn replied:
This computes the $T^{-1}(z)$ that appears in the contour integral in subspace methods. Adding a docstring :+1:

vincent-maillou replied:
perfect, thanks

vincent-maillou commented on src/qttools/obc/sancho_rubio.py:
Sancho-rubio is only using NumPy? Can it be made xp.? Is there a point in doing so? For consistency probably at least?

vetschn replied:
Don't know how i missed this, yes it should be `xp` of course. Will need to check why the tests pass on GPU as well :sweat_smile:

vincent-maillou replied:
perfect

vincent-maillou commented on src/qttools/obc/spectral.py:
```diff
@@ -0,0 +1,290 @@
+                v = vs[i][:, m]
+                w = ws[i, m]
+                # Direct computation of the surface Green's function.
+                inverse = npla.inv(v.conj().T @ a_ij[i] @ v * w)
```
x_ii is xp.zeros but the inverse is calling npla (that if I'm not wrong is NumPy), shouldn't it be xp.linalg? So it can invert on GPU too?

vetschn replied:
Yes, again something i overlooked. I actually also need to add tests that the different formulas work as expected.

vincent-maillou replied:
Sounds good

vincent-maillou reviewed (commented):
I left a bunch of comments, in general it's surely ready to be merged. Let me know when you've reviewed them and if nothing big comes out let me know, and I'll accept the merge. 

EDIT: Yes also I spotted that the coverage for this pull-request is quite low (68% against 78% of project). I know it's not the alpha-and-omega, what are your thoughts?

vetschn committed:
Put everything on GPU that can be put on GPU

I switched all the inversions to be GPU/CPU agnostic and made it a little bit more explicit when we actually need to call numpy.

I also added missing docstring and removed an unused method in the nevp.utils module.

Finally, i added tests for the memoizers and for the x_ii_formula kwarg in the Spectral solver.

vincent-maillou commented on tests/obc/conftest.py:
```diff
@@ -25,14 +27,20 @@
 
 @pytest.fixture(params=BLOCK_SIZE, autouse=True)
-def a_xx(request) -> np.ndarray:
+def a_xx(request) -> tuple[np.ndarray, ...]:
```
That's very minor but I think this should be xp. in the type-hint

vetschn replied:
Will address this in #47
